File: contests/lc/2141.MaximumRunningTimeofNComputers/solution.py
from typing import List


class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        # Example: n=2, batteries=[3,3,3]
        # total=9, hi=9//2=4. Check mid=3 -> feasible (sum min=6 >= 6).
        # Check mid=4 -> feasible (sum min=8 >= 8). Answer=4.
        #
        # Flow:
        # 1) Binary search on time in [0, total/n].
        # 2) A time t is feasible if sum(min(battery, t)) >= n * t.
        # 3) Return the largest feasible t.
        total = sum(batteries)
        lo, hi = 0, total // n

        def can_run(time: int) -> bool:
            energy = 0
            for charge in batteries:
                if charge >= time:
                    energy += time
                else:
                    energy += charge
            return energy >= n * time

        while lo < hi:
            mid = (lo + hi + 1) // 2
            if can_run(mid):
                lo = mid
            else:
                hi = mid - 1
        return lo

File: contests/lc/2141.MaximumRunningTimeofNComputers/solution.go
package maximumrunningtimeofncomputers

// maxRunTime returns the maximum minutes all n computers can run simultaneously.
// Uses binary search on time with a feasibility check based on total usable energy.
// Example: n=2, batteries=[3,3,3]
// total=9 -> hi=4. mid=2 (feasible), mid=3 (feasible), mid=4 (feasible) => answer=4.
//
// Flow:
// 1) Search time in [0, total/n].
// 2) time t is feasible when sum(min(battery, t)) >= n*t.
// 3) Keep the largest feasible t.
func maxRunTime(n int, batteries []int) int64 {
	var total int64
	for _, b := range batteries {
		total += int64(b)
	}

	lo, hi := int64(0), total/int64(n)

	canRun := func(t int64) bool {
		var energy int64
		for _, b := range batteries {
			if int64(b) >= t {
				energy += t
			} else {
				energy += int64(b)
			}
		}
		return energy >= t*int64(n)
	}

	for lo < hi {
		mid := (lo + hi + 1) / 2
		if canRun(mid) {
			lo = mid
		} else {
			hi = mid - 1
		}
	}
	return lo
}
